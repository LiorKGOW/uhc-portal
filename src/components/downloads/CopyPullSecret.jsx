import React from 'react';
import PropTypes from 'prop-types';
import { Button, Tooltip } from '@patternfly/react-core';
import { CopyIcon } from '@patternfly/react-icons/dist/esm/icons/copy-icon';
import { CopyToClipboard } from 'react-copy-to-clipboard';
import isEmpty from 'lodash/isEmpty';
import { trackEvents } from '~/common/analytics';
import withAnalytics from '~/hoc/withAnalytics';

class CopyPullSecret extends React.Component {
  constructor(props) {
    super(props);
    this.timer = null;
  }

  state = { clicked: false };

  onCopy = () => {
    if (this.timer) {
      window.clearTimeout(this.timer);
    }
    this.setState({ clicked: true }, () => {
      this.timer = window.setTimeout(() => {
        this.setState({ clicked: false });
        this.timer = null;
      }, 2500);
    });
  };

  render() {
    const { token, text, variant, pendoID, track } = this.props;
    const isDisabled = !token || !!token.error || isEmpty(token);
    const { clicked } = this.state;
    const tokenView = token.error ? '' : `${JSON.stringify(token)}\n`;

    const linkText = variant === 'link-inplace' && clicked ? 'Copied!' : text;

    const button = (
      <CopyToClipboard text={isDisabled ? '' : tokenView} onCopy={this.onCopy}>
        <Button
          variant="link"
          type="button"
          tabIndex={0}
          isAriaDisabled={isDisabled}
          icon={<CopyIcon />}
          onClick={() => {
            track(trackEvents.CopyPullSecret, { path: pendoID });
          }}
        >
          {linkText}
        </Button>
      </CopyToClipboard>
    );

    if (variant === 'link-inplace') {
      return button;
    }
    return (
      <Tooltip trigger="manual" content="Copied!" position="right" isVisible={clicked}>
        {button}
      </Tooltip>
    );
  }
}
CopyPullSecret.propTypes = {
  pendoID: PropTypes.string,
  track: PropTypes.object.isRequired,
  token: PropTypes.object.isRequired,
  text: PropTypes.string,
  variant: PropTypes.oneOf(['link-tooltip', 'link-inplace']).isRequired,
};
CopyPullSecret.defaultProps = {
  text: 'Copy pull secret',
};

export default withAnalytics(CopyPullSecret);
