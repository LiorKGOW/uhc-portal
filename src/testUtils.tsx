import React from 'react';
import { Provider } from 'react-redux';
import { createStore, Store } from 'redux';
import { render, RenderOptions } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import { toHaveNoViolations, axe } from 'jest-axe';

import { createBrowserHistory } from 'history';
import { store } from './redux/store';
import { reduxReducers } from './redux/reducers';
import * as restrictedEnv from './restrictedEnv';

const history = createBrowserHistory();

export const TestWrapper = ({
  children,
  initialStore,
}: {
  children: React.ReactNode;
  initialStore?: Store;
}) => <Provider store={initialStore || store}>{children}</Provider>;

const renderWithState = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
  initialState?: any, // setting to any because only a partial state structure may be sent
) => {
  const newStore = initialState ? createStore(reduxReducers(history), initialState) : store;

  const Wrapper = ({ children }: { children: React.ReactNode }) => (
    <TestWrapper initialStore={newStore}>{children}</TestWrapper>
  );
  return { ...render(ui, { wrapper: Wrapper, ...options }), user: userEvent.setup() };
};

export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';

// Type not exported in the library
export type UserEventType = {
  type: (
    field: HTMLElement,
    typeValue: string,
    { initialSelectionStart }: { initialSelectionStart: number },
  ) => void;
  clear: (field: HTMLElement) => void;
};

export { renderWithState as render };

/* ***** Items outside of React Test Library ************ */
expect.extend(toHaveNoViolations);

export const checkAccessibility = async (container: HTMLElement | string) => {
  const results = await axe(container);
  expect(results).toHaveNoViolations();
};

export const insightsMock = () => {
  global.insights = {
    chrome: {
      ...global.insights.chrome,
      on: () => () => {},
      appNavClick: () => {},
      auth: {
        getUser: () => Promise.resolve({ data: {} }),
        getToken: () => Promise.resolve(),
        getOfflineToken: () => Promise.resolve({ data: { refresh_token: 'hello' } }),
      },
    },
  };
};

export const mockRestrictedEnv = () => {
  const mock = jest.spyOn(restrictedEnv, 'isRestrictedEnv');
  mock.mockReturnValue(false);
  return mock;
};
