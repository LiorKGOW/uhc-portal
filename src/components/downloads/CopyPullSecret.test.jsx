import React from 'react';
import { screen, render, checkAccessibility } from '~/testUtils';

import CopyPullSecret from './CopyPullSecret';

const variants = ['link-tooltip', 'link-inplace'];

describe('<CopyPullSecret />', () => {
  const token = { auths: { foo: 'bar' } };
  const errorToken = { error: 'my error' };

  describe.each(variants)('with token and variant %s', (variant) => {
    it('is accessible', async () => {
      const { container } = render(<CopyPullSecret variant={variant} token={token} />);
      await checkAccessibility(container);
    });

    it('displays an enabled copy button', () => {
      render(<CopyPullSecret variant={variant} token={token} />);
      expect(screen.getByRole('button', { name: 'Copy pull secret' })).toHaveAttribute(
        'aria-disabled',
        'false',
      );
    });
  });

  describe.each(variants)('with error and variant %s', (variant) => {
    it('is accessible', async () => {
      const { container } = render(<CopyPullSecret variant={variant} token={errorToken} />);
      await checkAccessibility(container);
    });

    it('displays an disabled copy button', () => {
      render(<CopyPullSecret variant={variant} token={errorToken} />);
      expect(screen.getByRole('button', { name: 'Copy pull secret' })).toHaveAttribute(
        'aria-disabled',
        'true',
      );
    });
  });
});
