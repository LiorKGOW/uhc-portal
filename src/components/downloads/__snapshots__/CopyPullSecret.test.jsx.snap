// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CopyPullSecret /> with error link-inplace should render 1`] = `
<AnalyticsWrapper
  render={[Function]}
/>
`;

exports[`<CopyPullSecret /> with error link-tooltip should render 1`] = `
<Tooltip
  content="Copied!"
  isVisible={false}
  position="right"
  trigger="manual"
>
  <AnalyticsWrapper
    render={[Function]}
  />
</Tooltip>
`;

exports[`<CopyPullSecret /> with token link-inplace should render 1`] = `
<AnalyticsWrapper
  render={[Function]}
/>
`;

exports[`<CopyPullSecret /> with token link-tooltip should render 1`] = `
<Tooltip
  content="Copied!"
  isVisible={false}
  position="right"
  trigger="manual"
>
  <AnalyticsWrapper
    render={[Function]}
  />
</Tooltip>
`;
